core/trino-spi/src/main/java/io/trino/spi/type/Decimals.java

```diff
@@ -295,17 +297,44 @@ public static void writeShortDecimal(BlockBuilder blockBuilder, long value)
     public static long rescale(long value, int fromScale, int toScale)
     {
         if (toScale < fromScale) {
-            throw new IllegalArgumentException("target scale must be larger than source scale");
+            // Use MAX_SHORT_PRECISION to fit all precision long values
+            return shortToShortCast(
+                    value,
+                    MAX_SHORT_PRECISION,
+                    fromScale,
+                    MAX_SHORT_PRECISION,
+                    toScale,
+                    longTenToNth(fromScale - toScale),
+                    longTenToNth(fromScale - toScale - 1) * 5);
         }
-        return value * longTenToNth(toScale - fromScale);
+
+        long returnValue = value * longTenToNth(toScale - fromScale);
+
+        if (overflows(returnValue, MAX_SHORT_PRECISION)) {
+            throw new TrinoException(
+                    NUMERIC_VALUE_OUT_OF_RANGE,
+                    format("Rescaled value for %s is out of range", Decimals.toString(value, fromScale)));
+        }
+
```

Comment:
nit : Can we capture in a separate commit.

Reply:
Do we want separate commits for both overflow checks or should I merge them into a single commit?